### PR TITLE
Mlplan/feature/multilabel

### PR DESCRIPTION
**Pull request by mwever**
Refactored ML-Plan and its builder to support MLC with MEKA.

### CODE REVIEW

**Comment by mwever** [comment]
SonarCloud does fail as SQLAdapter needs to be refactored (see ticket AIL-92) and due to false positives because of assertions which do not have any effect in the compiled binaries, apparently wrong exception handling (although it is rethrown) and some code complexity issues with two many parameters etc.

File: JAICore/jaicore-basic/src/main/java/jaicore/basic/ListHelper.java [hunk 1]
```diff
@@ -5,36 +5,36 @@
 
 public class ListHelper {
```
Comment (fmohr) [inline JAICore/jaicore-basic/src/main/java/jaicore/basic/ListHelper.java]: This class looks admittedly strange to me. I think that this should be covered in SetUtil modulo renaming. Do you really want/need this for an `Iterable`?

File: JAICore/jaicore-basic/src/main/java/jaicore/basic/kvstore/KVStore.java [hunk 2]
```diff
@@ -266,26 +332,31 @@ public boolean matches(final Map<String, String> selection) {
...
 					currentValue = currentValue.split("#$#")[1];
 				} else {
 					matchesPattern = false;
-					break;
 				}
 
 				if (i == (exprSplit.length - 1) && !x.getValue().endsWith("*") && !currentValue.endsWith(exprSplit[i])) {
 					matchesPattern = false;
+				}
+
```
Comment (fmohr) [inline JAICore/jaicore-basic/src/main/java/jaicore/basic/kvstore/KVStore.java]: I know that this is SonarQube convention. But is this really more readable than with the breaks in place? We should discuss whether we want to keep this rule.

File: JAICore/jaicore-ml/src/main/java/jaicore/ml/core/evaluation/measure/ClassifierMetricGetter.java [hunk 3]
```diff
@@ -0,0 +1,238 @@
...
+import weka.classifiers.Evaluation;
+
+/**
+* Class for getting metrics by their name for single- and multilabel
+* classifiers.
+*
+* @author helegraf, mwever
+*
+*/
+public class ClassifierMetricGetter {
```
Comment (fmohr) [inline JAICore/jaicore-ml/src/main/java/jaicore/ml/core/evaluation/measure/ClassifierMetricGetter.java]: This class looks a bit weird to me. For simple multi-class, it is just a forwarder for an Evaluation object. Perhaps, the benefits of using this class could be explained in the class JavaDoc.

File: JAICore/jaicore-ml/src/main/java/jaicore/ml/core/evaluation/measure/EEvaluationMode.java [hunk 4]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.io.Serializable;
+
+/**
+* Represents evaluation modes. A mode can either be test or validation.
+*
+* @author helegraf
+*
+*/
+public enum EEvaluationMode implements Serializable {
```
Comment (fmohr) [inline JAICore/jaicore-ml/src/main/java/jaicore/ml/core/evaluation/measure/EEvaluationMode.java]: This also looks weird to me. Seems to be quite arbitrary.

File: JAICore/jaicore-ml/src/main/java/jaicore/ml/core/evaluation/measure/EventEmittingMeasure.java [hunk 5]
```diff
@@ -9,39 +9,55 @@
 import jaicore.ml.evaluation.MeasureListComputationEvent;
 import jaicore.ml.evaluation.MeasureSingleComputationEvent;
 
-public class EventEmittingMeasure<INPUT, OUTPUT> implements IMeasure<INPUT, OUTPUT> {
```
Comment (fmohr) [inline JAICore/jaicore-ml/src/main/java/jaicore/ml/core/evaluation/measure/EventEmittingMeasure.java]: I understand the idea here, but wouldn't it be more nautral that this is an interface? The problem I see here with using the decorator pattern is that all the interfaces implemented by the base measure get "lost".

My suggestion would be to have this an interface and use an abstract class to provide default implementations.

File: JAICore/jaicore-ml/src/main/java/jaicore/ml/core/evaluation/measure/multilabel/OverviewMeasureLoss.java [hunk 6]
```diff
@@ -0,0 +1,22 @@
+package jaicore.ml.core.evaluation.measure.multilabel;
+
+/**
+* Measure combining exact match, hamming loss, f1macroavgL and rankloss. Here
+* implemented in inverse.
+*
+* @author helegraf, mwever
+*
+*/
+public class OverviewMeasureLoss extends ADecomposableMultilabelMeasure {
```
Comment (fmohr) [inline JAICore/jaicore-ml/src/main/java/jaicore/ml/core/evaluation/measure/multilabel/OverviewMeasureLoss.java]: This is really a very strange measure (I believe). Even if you keep this, a name different from "Overview" would be a good thing. This is somehow a cross-metric "average" measure. I do not believe that this number has any reasonable semantic.

File: JAICore/jaicore-basic/src/main/java/jaicore/basic/ListHelper.java (hunk 1 above)
Comment (mwever) [reply JAICore/jaicore-basic/src/main/java/jaicore/basic/ListHelper.java]: Methods moved to SetUtil + Restricted Iterable to Collection.

**Pull request merged**